app/controllers/whats_app/webhook_controller.rb

```diff
@@ -0,0 +1,70 @@
+    UNSUCCESSFUL_DELIVERY = %w[undelivered failed].freeze
+
+    def message
+      adapter = WhatsAppAdapter::Inbound.new
+
+      adapter.on(WhatsAppAdapter::UNKNOWN_CONTRIBUTOR) do |whats_app_phone_number|
+        exception = WhatsAppAdapter::UnknownContributorError.new(whats_app_phone_number: whats_app_phone_number)
+        ErrorNotifier.report(exception)
+      end
+
+      adapter.on(WhatsAppAdapter::RESPONDING_TO_TEMPLATE_MESSAGE) do |contributor|
+        contributor.update!(whats_app_message_template_responded_at: Time.current, whats_app_template_message_sent_at: nil)
+        message = contributor.received_messages.first
+        WhatsAppAdapter::Outbound.send!(message)
+      end
+
+      adapter.on(WhatsAppAdapter::UNKNOWN_CONTENT) do |contributor|
+        WhatsAppAdapter::Outbound.send_unknown_content_message!(contributor)
+      end
+
+      whats_app_webhook_params = webhook_params.to_h.transform_keys(&:underscore)
+      adapter.consume(whats_app_webhook_params) { |message| message.contributor.reply(adapter) }
+    end
+
+    def errors
+      return unless webhook_params['Level'] == 'ERROR' && JSON.parse(params['Payload'])['error_code'].to_i == 21_617
+
+      payload = JSON.parse(webhook_params['Payload'])
+      more_info = payload['more_info']
+      parameters = payload['webhook']['request']['parameters']
+      error_text = more_info['Msg'].gsub(Setting.twilio_account_sid, '<ACCOUNT_SID>')
+      exception = WhatsAppAdapter::MaximumMessageLengthExceededError.new(text: error_text,
+                                                                         server_phone_number: parameters['channelToAddress'])
+      ErrorNotifier.report(exception, context: { error_sid: webhook_params['Sid'], message_sid: parameters['messageSid'] })
+    end
+
+    def status
+      return unless status_params['MessageStatus'].in?(UNSUCCESSFUL_DELIVERY)
+
+      exception = WhatsAppAdapter::MessageDeliveryUnsuccessfulError.new(status: status_params['MessageStatus'],
+                                                                        whats_app_phone_number: status_params['To'].split('whatsapp:').last)
+      ErrorNotifier.report(exception, context: { message_sid: status_params['MessageSid'] })
+    end
+
+    private
+
+    def webhook_params
+      params.permit(:AccountSid, :ApiVersion, :Body, :From, :Level, :Latitude, :Longitude, :MessageSid, :NumMedia, :NumSegments,
+                    :ParentAccountSid, :Payload, :PayloadType, :ProfileName, :ReferralNumMedia, :Sid, :SmsMessageSid, :SmsSid,
+                    :SmsStatus, :Timestamp, :To, :WaId, *media_params)
+    end
+
+    def status_params
+      params.permit(:AccountSid, :ApiVersion, :ChannelInstallSid, :ChannelPrefix, :ChannelToAddress, :ErrorCode, :EventType,
+                    :From, :MessageSid, :MessageStatus, :SmsSid, :SmsStatus, :To)
+    end
+
+    def media_params
+      indeces_of_media = params[:NumMedia].to_i
```

Comment:
I think plural of index is "indices".

Reply:
For what is `media_params` being used? I don't see it.

Reply:
> I think plural of index is "indices".

yeah, haha I guess as a native speaker, I should know that. I mixed `indexes` and `indices`

> For what is media_params being used? I don't see it.

It is used in the strong params to permit `MediaContentTypeX/MediaUrlX`, where `X` is the index of the file.

[Twilio docs](https://www.twilio.com/docs/whatsapp/tutorial/send-and-receive-media-messages-whatsapp-ruby#get-urls-to-the-media) say:

> Twilio sends form variables named MediaUrlX, where X is a zero-based index.

But, I actually made this more complex than it need be since it also says:

> WhatsApp messages will only contain one media file per incoming message, so you can access the file at MediaUrl0 on the incoming request from Twilio to your webhook URL.

This was also my experience with sending multiple images. I was already considering refactoring it to simplify the logic, but now that I read this closer, I think it's definitely worth it.